pkg/cloudprovider/vsphereparavirtual/route.go

```diff
@@ -0,0 +1,353 @@
+	}
+	return r.createCPRoutes(routeSets), nil
+}
+
+// createCPRoutes creates cloudprovider Routes based on RouteSet CR
+func (r *routesProvider) createCPRoutes(routeSets *v1alpha1.RouteSetList) []*cloudprovider.Route {
+	var routes []*cloudprovider.Route
+	for _, routeSet := range routeSets.Items {
+		// only return cloudprovider.Route if RouteSet CR status 'Ready' is true
+		condition := GetRouteSetCondition(&(routeSet.Status), v1alpha1.RouteSetConditionTypeReady)
+		if condition != nil && condition.Status == v1.ConditionTrue {
+			// one RouteSet per node, so we can use nodeName as the name of RouteSet CR
+			nodeName := routeSet.Name
+			for _, route := range routeSet.Spec.Routes {
+				cpRoute := &cloudprovider.Route{
+					Name:            route.Name,
+					TargetNode:      types.NodeName(nodeName),
+					DestinationCIDR: route.Destination,
+				}
+				routes = append(routes, cpRoute)
+			}
+		}
+	}
+	return routes
+}
+
+// CreateRoute implements Routes.CreateRoute
+// Create a RouteSet custom resource for a Node
+func (r *routesProvider) CreateRoute(ctx context.Context, clusterName string, nameHint string, route *cloudprovider.Route) error {
+	nodeName := string(route.TargetNode)
+	klog.V(6).Infof("Creating Route for node %s with hint %s in cluster %s", nodeName, nameHint, clusterName)
+
+	nodeIP, err := r.getNodeIPAddress(nodeName, util.IsIPv4(route.DestinationCIDR))
+	if err != nil {
+		klog.Errorf("getting node %s IP address failed: %v", nodeName, err)
+		return err
+	}
+
+	routeSet, err := r.createRouteSetCR(ctx, clusterName, nameHint, nodeName, route.DestinationCIDR, nodeIP)
+	if err != nil {
+		klog.Errorf("creating RouteSet CR for node %s failed: %s", nodeName, err)
+		return err
+	}
+	// check realized state of static routes
+	return r.checkStaticRouteRealizedState(routeSet.Name)
+}
+
+// createRouteSetCR creates RouteSet CR through RouteSet client
+func (r *routesProvider) createRouteSetCR(ctx context.Context, clusterName string, nameHint string, nodeName string, cidr string, nodeIP string) (*v1alpha1.RouteSet, error) {
+	labels := map[string]string{
+		LabelKeyClusterName: clusterName,
+	}
+	nodeRef := metav1.OwnerReference{
+		APIVersion: "v1",
+		Kind:       "Node",
+		Name:       nodeName,
+		UID:        types.UID(nameHint),
+	}
+	owners := make([]metav1.OwnerReference, len(r.ownerRefs))
+	copy(owners, r.ownerRefs)
+	owners = append(owners, nodeRef)
+	route := v1alpha1.Route{
+		Name:        r.GetRouteName(nodeName, cidr, clusterName),
+		Destination: cidr,
+		Target:      nodeIP,
+	}
+	routeSetSpec := v1alpha1.RouteSetSpec{
+		Routes: []v1alpha1.Route{
+			route,
+		},
+	}
+	routeSet := &v1alpha1.RouteSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            nodeName,
+			OwnerReferences: owners,
+			Namespace:       r.namespace,
+			Labels:          labels,
+		},
+		Spec: routeSetSpec,
+	}
+
+	_, err := r.routeClient.NsxV1alpha1().RouteSets(r.namespace).Create(ctx, routeSet, metav1.CreateOptions{})
```

Comment:
If the RouteSet CR is already created, will there be any error here? Or it just apply to the existing CR

Reply:
yes, the error will be caught and `ErrCreateRouteSet` will be returned, in that case, we won't create a duplicate CR

Reply:
Do you think we should not throw `ErrCreateRouteSet` if the CR is already created?  especially in cloud provider restart case, maybe it is a little confusing to report these errors.

Reply:
but I think if the routeSet CR already exists, maybe we should not regard it as an error, we just do nothing afterwards

Reply:
I agree, it should not be an error.